### PR TITLE
Implement message components

### LINKED ISSUE #624: Implement message components
https://discord.com/developers/docs/interactions/message-components

This must be implemented in a way that is extendable to further components without creating breaking changes

### PR DESCRIPTION
**Pull request by FasterSpeeding**
### Summary
Implement message components

### Checklist

- [x] I have run `nox` and all the pipelines have passed.
- [x] I have made unittests according to the code I have added/modified/deleted.

### Related issues
Closes #624
~~This is blocked by #701~~

### CODE REVIEW

File: hikari/api/rest.py
Comment (FasterSpeeding) [inline hikari/api/rest.py:6826]: what about just `build_action_row`?

File: hikari/impl/special_endpoints.py
Comment (davfsa) [inline hikari/impl/special_endpoints.py:991]: Are these necessary or can they follow a creation pattern like embed does?

Note, I completely missed to ask the same for the other two.

File: hikari/interactions/components.py
Comment (FasterSpeeding) [reply hikari/interactions/components.py:274]: the name shouldn't matter that much since that argument is positional only

File: hikari/impl/special_endpoints.py
Comment (FasterSpeeding) [reply hikari/impl/special_endpoints.py:991]: These builders don't match up with their relevant API representations and merging the two would infer that the received objects are mutable when they aren't intended to be and make their typing worse/less concise.

File: hikari/interactions/components.py
Comment (davfsa) [reply hikari/interactions/components.py:274]: What about in the docstring?

File: hikari/impl/special_endpoints.py
Comment (davfsa) [reply hikari/impl/special_endpoints.py:991]: As far as I can see, they seem to be the same but with some defaults, which can be implemented into the objects

Comment (FasterSpeeding) [reply hikari/impl/special_endpoints.py]: id is optional on command builder unlike Command where it is not
Button builder only takes a custom emoji ID whereas Button will have a proper custom emoji object (id and name and whether it's animated)

Comment (davfsa) [reply hikari/impl/special_endpoints.py]: This can be managed by the deserializing and serializing of the model

Comment (FasterSpeeding) [reply hikari/impl/special_endpoints.py]: So you want to make Command.id optional?

**Comment by FasterSpeeding** [comment]
Gonna switch over to a nested builder pattern for handling the components builders

**Comment by FasterSpeeding** [comment]
Gonna add select menus into this PR instead of waiting for it to be reviewed and merged

**Comment by FasterSpeeding** [comment]
This is blocked by #701

File: hikari/api/entity_factory.py
Comment (FasterSpeeding) [inline hikari/api/entity_factory.py:1170]: Wrong type

File: hikari/messages.py
Comment (FasterSpeeding) [inline hikari/messages.py:677]: this is now 100

File: hikari/messages.py
Comment (FasterSpeeding) [inline hikari/messages.py:683]: this is now 100

**Comment by FasterSpeeding** [comment]
- interactions on ephemeral messages will now include the entire message payload, instead of just the id and flags.

**Comment by FasterSpeeding** [comment]
bump

**Comment by NotNite** [comment]
*bump*

any ETA for this to be merged? it's blocking development of my bot

**Comment by davfsa** [comment]
> *bump*
> 
> any ETA for this to be merged? it's blocking development of my bot

@NotNite I am currently on holiday and reviewing such huge PR without a proper testing environment and PC is extremely hard. I will probably be reviewing it in the new week and it might be merged short after

File: hikari/impl/special_endpoints.py
Comment (davfsa) [inline hikari/impl/special_endpoints.py:956]:
```suggestion
    typing.Tuple[hikari.undefined.UndefinedOr[builtins.str], hikari.undefined.UndefinedOr[builtins.str]]
```

**Review by davfsa — APPROVED** [review]

**Review by davfsa — APPROVED** [review]

**Comment by FasterSpeeding** [comment]
@davfsa you can merge this whenever if you think it's ready

**Pull request merged**
